src/test/java/com/amazon/dlic/auth/http/jwt/keybyoidc/SelfRefreshingKeySetTest.java

```diff
@@ -11,66 +11,91 @@
     }
 
-    @Test(timeout = 10000)
-    public void twoThreadedTest() throws Exception {
-        BlockingMockKeySetProvider provider = new BlockingMockKeySetProvider();
+    @Test
+    public void getKey_withKidShouldReturnValidKey() throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        final SelfRefreshingKeySet selfRefreshingKeySet = new SelfRefreshingKeySet(provider);
+        OctetSequenceKey key = (OctetSequenceKey) selfRefreshingKeySet.getKey("kid/a");
+        assertThat(keyForKidA, is(equalTo(key.getKeyValue().decodeToString())));
+    }
 
-        ExecutorService executorService = Executors.newCachedThreadPool();
+    @Test
+    public void getKey__withNullOrInvalidKidShouldThrowAnException() throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        Future<JWK> f1 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/a"));
+        Assert.assertThrows(AuthenticatorUnavailableException.class, () -> selfRefreshingKeySet.getKey(null));
+        Assert.assertThrows(BadCredentialsException.class, () -> selfRefreshingKeySet.getKey("kid/X"));
+    }
 
-        provider.waitForCalled();
+    @Test
+    public void getKeyAfterRefresh_withKidShouldReturnKey() throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        Future<JWK> f2 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/b"));
+        OctetSequenceKey key = (OctetSequenceKey) selfRefreshingKeySet.getKeyAfterRefresh("kid/b");
+        assertThat(keyForKidB, is(equalTo(key.getKeyValue().decodeToString())));
+    }
 
-        while (selfRefreshingKeySet.getQueuedGetCount() == 0) {
-            Thread.sleep(10);
+    @Test
+    public void getKeyAfterRefresh_withMultipleCallsShouldIncreaseQueueCount() throws InterruptedException, ExecutionException {
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+        String[] keys = new String[] { "kid/a", "kid/b" };
+        for (int i = 0; i < numThreads; i++) {
+            // Using executor to make multiple asynchronous calls to method getKeyAfterRefresh, so queuedGetCount gets increased.
+            // Without executor block, getKeyAfterRefresh method would be called once on each iteration in the main thread and wait for the
+            // task to complete before continuing the loop, so queuedGetCount would not have pending tasks.
+            executor.execute(() -> {
+                try {
+                    int indexKey = (int) (Math.random() * 2);
+                    String keyToCompare = indexKey == 0 ? keyForKidA : keyForKidB;
+                    OctetSequenceKey key = (OctetSequenceKey) selfRefreshingKeySet.getKeyAfterRefresh(keys[indexKey]);
+
+                    assertThat(key, is(notNullValue()));
+                    assertThat(keyToCompare, is(equalTo(key.getKeyValue().decodeToString())));
+                } catch (AuthenticatorUnavailableException e) {} catch (BadCredentialsException e) {}
```

Comment:
we should add fail statements here

catch (AuthenticatorUnavailableException | BadCredentialsException e) {
    fail("No exception was expected but found: " + e.getMessage());
}